### PR TITLE
WIP - Add order_schema_markup action to output Schema markup for order emails.

### PR DESCRIPTION
**Pull request by jonkoops**
This is a work in progress pull request to implement Schema markup in order emails. For more information see the related issue #9745.

### CODE REVIEW

File: includes/class-wc-emails.php [hunk 1]
```diff
@@ -269,6 +270,108 @@ public function order_details( $order, $sent_to_admin = false, $plain_text = fal
...
+
+		switch ( $order->get_status() ) {
+			case 'pending':
+				$markup['orderStatus'] = 'http://schema.org/OrderPaymentDue';
+				break;
+			case 'processing':
+				$markup['orderStatus'] = 'http://schema.org/OrderProcessing';
+				break;
+			case 'on-hold':
+				$markup['orderStatus'] = 'http://schema.org/OrderProcessing';
```
Comment (jonkoops) [inline includes/class-wc-emails.php]: Not sure if OrderProcessing is the correct type for the 'On Hold' status. A complete list of possible order statuses can be found here: https://schema.org/OrderStatus.

**Comment by bentasm1** [comment]
Looks great!

File: includes/class-wc-emails.php (hunk 1 above)
Comment (mikejolley) [inline includes/class-wc-emails.php]: On-hold is awaiting manual review (to confirm payment, fix a problem etc). I wonder if OrderProblem is suitable?

Comment (jonkoops) [inline includes/class-wc-emails.php]: Yeah, I think that OrderProblem probably best describes this situation.

Comment (jonkoops) [inline includes/class-wc-emails.php]: Okay, changed it. I also replaced a duplicate `refunded` status with the `failed` status.

**Comment by mikejolley** [comment]
Code looks good. @mattyza since you reported this initially, did you want to do some testing? :)

**Comment by mattyza** [comment]
@mikejolley - Sure, will do. :+1:

**Comment by mikejolley** [comment]
@jonkoops did you actually get this showing in gmail? I've been unsuccessful. Sending to myself.

**Comment by bentasm1** [comment]
@mikejolley Google has to approve your account for the Schema to actually show up in the emails.  They won't just "work" because the code for them is there.  It's a two part process, requiring gmail's manual human approval.

**Comment by mikejolley** [comment]
@bentasm1 the docs say it works if you email yourself.

**Comment by justinshreve** [comment]
@mikejolley I couldn't get the self email to work either. From some searching, it looks like the self emails still need DKIM/SPF signing:

https://support.google.com/mail/answer/180707?hl=en# & https://developers.google.com/gmail/markup/registering-with-google (It mentions signing being a requirement for registration, but some StackOverflow threads say it is required for the self emails as well)

I don't have a domain/setup to easily test that at the moment... however I tested using this method https://developers.google.com/gmail/markup/apps-script-tutorial#creating_the_project and copying the markup from the order email (Select "Message text garbled?") from the email that WooCommerce sends after applying this branch.

@jonkoops It didn't work immediately for me. I checked why using the validator (https://www.google.com/webmasters/markup-tester/u/0/) and it returns:

```
price: missing and required
priceCurrency: missing and required
```

So we need the full order price & currency, in addition to the line item prices.

I added those in and tested the markup again and it worked.

File: includes/class-wc-emails.php [hunk 2]
```diff
@@ -269,6 +270,108 @@ public function order_details( $order, $sent_to_admin = false, $plain_text = fal
...
+				$item_offered['url'] = get_permalink( $product->get_id() );
+			}
+
+			if ( $image_id = $product->get_image_id() ) {
+				$item_offered['image'] = wp_get_attachment_image_url( $image_id, 'thumbnail' );
+			}
+
+			$accepted_offer = array(
+				'@type' => 'Offer',
+				'itemOffered' => $item_offered,
```
Comment (justinshreve) [inline includes/class-wc-emails.php]: All the new arrays introduced by your PR should have each value indented so they align. See https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#indentation for more information.

File: includes/class-wc-emails.php [hunk 3]
```diff
@@ -269,6 +270,108 @@ public function order_details( $order, $sent_to_admin = false, $plain_text = fal
...
+		$markup = array(
+			'@context' => 'http://schema.org',
+			'@type' => 'Order',
+			'merchant' => array(
+				'@type' => 'Organization',
+				'name' => get_bloginfo( 'name' )
+			),
+			'orderNumber' => strval( $order->get_order_number() ),
+			'acceptedOffer' => $accepted_offers,
+			'url' => $order->get_view_order_url()
```
Comment (justinshreve) [inline includes/class-wc-emails.php]: The last value in arrays should still have a trailing comma.

**Comment by mikejolley** [comment]
SUCCESS. Found an app to test these out.

**Comment by jonkoops** [comment]
@mikejolley I never got it to work with the actual e-mails, but I have used the tools provided by Google to validate the output.

@justinshreve I believe the properties `price` and `priceCurrency` on the `Order` itself are no longer part of the [Order schema markup](https://schema.org/Order). Maybe the markup tester is out of date? Could you perhaps share the output so I can verify this is the case?

**Comment by justinshreve** [comment]
@jonkoops It seems like the validator might be out of date then. Especially if it's missing from the schema and it worked for Mike without it.

The output generated code was

```
<script type="application/ld+json">{"@context":"http:\/\/schema.org","@type":"Order","merchant":{"@type":"Organization","name":"Woo"},"orderNumber":"12","acceptedOffer":[{"@type":"Offer","itemOffered":{"@type":"Product","name":"Park
City 2015
Waterbottle","url":"https:\/\/woo.dev\/product\/park-city-2015-waterbottle\/"},"price":15,"priceCurrency":"USD","eligibleQuantity":{"@type":"QuantitativeValue","value":"1"}}],"url":"https:\/\/woo.dev\/my-account\/view-order\/12","orderStatus":"http:\/\/schema.org\/OrderProcessing"}</script>
```

And the validator returned

screen shot 2016-03-30 at 11 05 42 am

**Comment by mikejolley** [comment]
I added those before merging. The nesting in that tool is confusing.

**Comment by jonkoops** [comment]
Still strange that these properties are present only in Google's testing tool and not the Shema.org specification. Guess it doesn't hurt to add them for Gmail even if it is off spec.

**Pull request merged**